### PR TITLE
Rule S1696: 'NullReferenceException' should not be caught

### LINKED ISSUE #583: Rule S1696: 'NullReferenceException' should not be caught
[RSPEC-1696](https://jira.sonarsource.com/browse/RSPEC-1696)

`NullReferenceException` should be avoided, not caught. Any situation in which `NullReferenceException` is explicitly caught can easily be converted to a null test, and any behavior being carried out in the catch block can easily be moved to the "is null" branch of the conditional.

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #583

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/RspecStrings.resx [hunk 1]
```diff
@@ -26,36 +26,36 @@
         <value>[base64 mime encoded string representing a byte array form of the .NET Framework object]</value>
         <comment>This is a comment</comment>
     </data>
-
-    There are any number of "resheader" rows that contain simple
+
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/RspecStrings.resx]: Exrtra spaces.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+using System;
+
+namespace Tests.Diagnostics
+{
+    class Program
+    {
+        int Foo(string s)
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs]: We may want to also support exception filters, e.g: 
`catch (Exception e) when (e is NullReferenceException)`
But it seems a very unlikely case, so if it's not trivial, it's fine to skip it.

But we need test case documenting the behaviour.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCatchNullReferenceException.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotCatchNullReferenceException : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S1696";
+        private const string MessageFormat = "Make the dereference conditional on its not being null.";
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCatchNullReferenceException.cs]: The message is weird. What about "Do not catch NullReferenceException; test for null instead"?

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs [hunk 4]
```diff
@@ -0,0 +1,20 @@
+using System;
```
Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs]: Checking for null is not always possible - for example a 3rd party library may be throwing.
Do we want to allow cases when exception is being caught, logged and rethrown?

Regardless of decision, need to add a test for that.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCatchNullReferenceException.cs (hunk 3 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCatchNullReferenceException.cs]: I am happy to change it, it also felt weird for me but I put that on my english level. I took the message from Java RSPEC cause I was a bit out of ideas.

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs (hunk 2 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs]: I haven't think of this one!

File: sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs (hunk 4 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs]: I was actually wondering for those cases and I thought that in this case the user will mark the issue as `won't fix` because they know it is fine in this case. WDYT?

Comment (michalb-sonar) [inline sonaranalyzer-dotnet/src/Tests/SonarAnalyzer.UnitTest/TestCases/DoNotCatchNullReferenceException.cs]: You're probably right, it's an edge case. So just add a test. Perhaps also add a test with nothing but throw (and throw e;)?
